R/slice.R

```diff
@@ -174,62 +174,129 @@ vec_assign_fallback <- function(x, i, value) {
+  maybe_get(vec_maybe_coerce_index(i))
+}
+vec_is_index <- function(i) {
+  maybe <- vec_maybe_coerce_index(i)
+  is_null(maybe$error)
+}
+
+vec_coerce_position <- function(i) {
+  maybe_get(vec_maybe_coerce_position(i))
+}
+vec_is_position <- function(i) {
+  maybe <- vec_maybe_coerce_position(i)
+  is_null(maybe$error)
+}
+
+vec_maybe_coerce_index <- function(i) {
   if (!vec_is(i)) {
-    stop_position_bad_type(i)
+    return(maybe(error = new_error_index_bad_type(i)))
   }
   if (is.object(i)) {
-    if (vec_is_subtype(i, lgl())) {
-      stop_position_bad_type(i)
-    } else if (vec_is_subtype(i, int())) {
+    if (vec_is_subtype(i, int())) {
       i <- vec_cast(i, int())
     } else if (vec_is_subtype(i, chr())) {
       i <- vec_cast(i, chr())
     } else {
-      stop_position_bad_type(i)
+      return(maybe(error = new_error_index_bad_type(i)))
     }
   } else if (is_double(i)) {
     i <- vec_coercible_cast(i, int())
   }
 
-  type <- typeof(i)
-  if (!type %in% c("integer", "character") ||
-      length(i) != 1L ||
+  if (!typeof(i) %in% c("integer", "character", "logical")) {
+    return(maybe(error = new_error_index_bad_type(i)))
+  }
+
+  maybe(i)
+}
+
+vec_maybe_coerce_position <- function(i) {
+  if (is.object(i) && vec_is(i) && vec_is_subtype(i, lgl())) {
+    return(maybe(error = new_error_position_bad_type(i)))
+  }
+
+  maybe <- vec_maybe_coerce_index(i)
+
+  # Return a subclass of index error
+  if (!is_null(maybe$error)) {
+    maybe$error <- new_error_position_bad_type(i = maybe$error$i)
+    return(maybe)
+  }
+
+  i <- maybe$value
+
+  if (typeof(i) == "logical") {
+    return(maybe(error = new_error_position_bad_type(i)))
+  }
+
+  # Positions must be size 1, can't be NA, and must be positive
+  if (length(i) != 1L ||
       is.na(i) ||
-      (type == "integer" && i < 1L)) {
-    stop_position_bad_type(i)
+      (typeof(i) == "integer" && i < 1L)) {
+    return(maybe(error = new_error_position_bad_type(i)))
   }
 
-  vec_as_index(i, n, names = names)
+  maybe
 }
 
-stop_position_bad_type <- function(i) {
-  # Should we derive from `stop_incompatible_type()` once we have
-  # union types? The index is incompatible with `union<chr(), int()>`.
-  abort("", "vctrs_error_position_bad_type", i = i)
+new_error_index_bad_type <- function(i, ..., .subclass = NULL) {
+  error_cnd(
+    c(.subclass, "vctrs_error_index_bad_type"),
+    i = i,
+    ...
+  )
+}
+new_error_position_bad_type <- function(i, ..., .subclass = NULL) {
+  new_error_index_bad_type(
+    .subclass = c(.subclass, "vctrs_error_position_bad_type"),
+    i = i,
+    ...
+  )
 }
 
 #' @export
-conditionMessage.vctrs_error_position_bad_type <- function(c) {
+conditionMessage.vctrs_error_index_bad_type <- function(c) {
   i <- c$i
   arg <- c$arg %||% "i"
+  issue <- "Must subset with positions or names."
 
-  lead <- "Must extract with a single position or name."
+  if (is.object(i) || !typeof(i) %in% c("integer", "character", "logical")) {
```

Comment:
It feels wrong to me to have to repeat this logic in the error message generator. If you we have to  generate this message lazily, maybe it's better to create a closure in `vec_maybe_coerce_index()`?

Reply:
That's a great idea, I'll try that next, outside of this PR.